packages/connection-manager/src/ConnectionManager.js

```diff
@@ -0,0 +1,168 @@
+       */
+      instance: this,
+
+      /**
+       * @property {ConnectionQueue} waitingConnections
+       * @description List of waiting connections ordered by priority
+       * @name ConnectionManager~Context#waitingConnections
+       */
+      waitingConnections: new ConnectionQueue(),
+
+      /**
+       * @property {List} openConnections
+       * @description List of open connections
+       * @name ConnectionManager~Context#next
+       */
+      openConnections: new ConnectionQueue(),
+
+      /**
+       * @property {function} next
+       * @description Opens the the connection if there's a free slot.
+       * @name ConnectionManager~Context#next
+       */
+      next() {
+        if (
+          context.openConnections.size >= maxConnections ||
+          context.waitingConnections.size === 0
+        ) {
+          return;
+        }
+
+        const connection = context.waitingConnections.first();
+
+        if (connection.state === AbstractConnection.INIT) {
+          connection.open(connection.url);
+        }
+
+        if (connection.state === AbstractConnection.OPEN) {
+          context.openConnections = context.openConnections.enqueue(connection);
+        }
+
+        // after added to open list, we can remove it from waiting
+        context.waitingConnections = context.waitingConnections.shift();
+        context.next();
+      },
+
+      /**
+       * @property {function} handleConnectionAbort
+       * @name ConnectionManager~Context#handleConnectionAbort
+       * @param {ConnectionEvent} event
+       */
+      handleConnectionAbort: event => {
+        this.dequeue(event.target);
+      },
+
+      /**
+       * @property {function} handleConnectionComplete
+       * @name ConnectionManager~Context#handleConnectionComplete
+       * @param {ConnectionEvent} event
+       */
+      handleConnectionComplete: event => {
+        this.dequeue(event.target);
+      },
+
+      /**
+       * @property {function} handleConnectionError
+       * @name ConnectionManager~Context#handleConnectionError
+       * @param {ConnectionEvent} event
+       */
+      handleConnectionError: event => {
+        this.dequeue(event.target);
+      }
+    };
+
+    this.enqueue = this.enqueue.bind(context);
+    this.dequeue = this.dequeue.bind(context);
+  }
+
+  /**
+   * Queues given connection with given priority
+   *
+   * @this ConnectionManager~Context
+   * @param {AbstractConnection} connection – connection to queue
+   * @param {Integer} [priority] – optional change of priority
+   */
+  enqueue(connection, priority) {
+    if (connection.state === AbstractConnection.CLOSED) {
+      return;
+    }
+
+    if (connection.state === AbstractConnection.INIT) {
+      this.waitingConnections = this.waitingConnections.enqueue(
+        connection,
+        priority
+      );
+    }
+
+    if (connection.state === AbstractConnection.OPEN) {
+      this.openConnections = this.openConnections.enqueue(connection);
```

Comment:
So we can just open all the connections upfront and they go directly into `openConnections` skipping `maxConnections` check :trollface:?

Reply:
Yes, because if the connection is open, it is open.. it doesnt matter if it shouldnt be opened or not.